calliope/core/util/observed_dict.py

```diff
@@ -0,0 +1,73 @@
+
+"""
+Copyright (C) 2013-2019 Calliope contributors listed in AUTHORS.
+Licensed under the Apache 2.0 License (see LICENSE file).
+
+"""
+
+from collections import Mapping
+from calliope.core.attrdict import AttrDict
+
+
+class ObservedDict(dict):
+    """
+    Dictionary subclass which updates an observer when there is a change in the
+    values assigned to keys in the Dictionary.
```

Comment:
This creates a __copy__ of the initial_dict, right? And it can also take YAML string which it roundtrips through AttrDict, but that isn't mentioned in the docstring, and I'm not sure if it's a good idea (either have clear and tight coupling to AttrDict, or leave this as a pure-dict implementation)?

Reply:
yes, a **copy** of initial_dict. Agreed on it not being clean to have the AttrDict YAML -> dict within it. But AttrDict is already sullying `UpdateObserver`, so I didn't see the harm (provided it was documented, ofc).